zk/stages/stage_batches.go

```diff
@@ -76,26 +84,48 @@ type HermezDb interface {
 type DatastreamClient interface {
 	ReadAllEntriesToChannel() error
 	GetEntryChan() chan interface{}
+	GetL2BlockByNumber(blockNum uint64) (*types.FullL2Block, error)
+	GetLatestL2Block() (*types.FullL2Block, error)
 	GetLastWrittenTimeAtomic() *atomic.Int64
 	GetStreamingAtomic() *atomic.Bool
 	GetProgressAtomic() *atomic.Uint64
 	EnsureConnected() (bool, error)
+	Start() error
+	Stop()
 }
 
+type dsClientCreatorHandler func(context.Context, *ethconfig.Zk, uint64) (DatastreamClient, error)
+
 type BatchesCfg struct {
-	db                  kv.RwDB
-	blockRoutineStarted bool
-	dsClient            DatastreamClient
-	zkCfg               *ethconfig.Zk
+	db                   kv.RwDB
+	blockRoutineStarted  bool
+	dsClient             DatastreamClient
+	dsQueryClientCreator dsClientCreatorHandler
+	zkCfg                *ethconfig.Zk
 }
 
-func StageBatchesCfg(db kv.RwDB, dsClient DatastreamClient, zkCfg *ethconfig.Zk) BatchesCfg {
-	return BatchesCfg{
+func StageBatchesCfg(db kv.RwDB, dsClient DatastreamClient, zkCfg *ethconfig.Zk, options ...Option) BatchesCfg {
```

Comment:
I'd prefer using explicit parameter naming here, since it is not a convention to use options as parameters in erigon.

Reply:
It is not the naming here, but instead it is the `Options` pattern and it compensates lack of the constructor/function overloading in the Golang.

In this specific use case, for the testing purposes, I need to provide a callback that would create the mock data stream client, i.e. `TestDataStreamClient` (instead of instantiating the one used outside of the tests).

So in the tests, I'm providing the callback that would allow me to create the mock implementation of DS client https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/zk/stages/stage_batches_test.go#L55. But since this is an optional parameter, the rest of the API is not affected https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/turbo/stages/zk_stages.go#L51 and that's the beauty of this approach.

I don't like providing an explicit parameter for data stream client creator callback function, because it is needed only for the testing purposes. However, if you insist to do so, or have a different proposal, I'm keen to listen out. 🙂 

Resources:
- https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c
- https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
Thanks for the context, it is a beautiful design. I am fine with keeping it.